From 4a9678bba382b0d69cf206a2933dbcd801c78e8f Mon Sep 17 00:00:00 2001
From: holishing <holishing@ccns.ncku.edu.tw>
Date: Sun, 2 Nov 2025 03:11:25 +0800
Subject: [PATCH] feat: try strlcpy/strlcat patch for cpp mode

---
 CMakeLists.txt | 13 +++++++++++++
 include/dao.h  |  4 ++++
 lib/string.c   |  4 ++++
 3 files changed, 21 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index b0af93164..fd4fca6eb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -22,6 +22,19 @@ include(CheckCCompilerFlag)
 include(CheckSymbolExists)
 include(FindPkgConfig)
 
+#
+
+set(CMAKE_REQUIRED_INCLUDES "string.h")
+check_symbol_exists(strlcpy "string.h" HAVE_STRLCPY)
+check_symbol_exists(strlcat "string.h" HAVE_STRLCAT)
+
+if (HAVE_STRLCPY)
+  add_definitions(-DHAVE_STRLCPY=1)
+endif()
+if (HAVE_STRLCAT)
+  add_definitions(-DHAVE_STRLCAT=1)
+endif()
+
 ## Toolchain settings
 
 if(NOT USE_CXX)
diff --git a/include/dao.h b/include/dao.h
index 514b7d036..18d9c1a94 100644
--- a/include/dao.h
+++ b/include/dao.h
@@ -189,8 +189,12 @@ GCC_NONNULLS void str_rtrim(char *buf);
 GCC_NONNULL(1) GCC_RET_NONNULL char *str_ttl_hdrmode(const char *title, enum HdrMode *pmode);
 GCC_NONNULLS GCC_PURE GCC_RET_NONNULL char *str_ttl(const char *title);
 GCC_NONNULLS void str_xor(char *dst GCC_NONSTRING, const char *src);
+#ifndef HAVE_STRLCAT
 GCC_NONNULLS size_t strlcat(char *dst, const char *src, size_t siz);
+#endif
+#ifndef HAVE_STRLCPY
 GCC_NONNULLS size_t strlcpy(char *dst, const char *src, size_t siz);
+#endif
 /* date.c */
 void str_stamp(char *str, const time_t *chrono);
 char *Btime(const time_t *clock);
diff --git a/lib/string.c b/lib/string.c
index b80f1e154..c6ac8daa9 100644
--- a/lib/string.c
+++ b/lib/string.c
@@ -1046,6 +1046,7 @@ GCC_NONNULLS
     }
 }
 
+#ifndef HAVE_STRLCAT
 /* strlcat based on OpenBSDs strlcat */
 
 /*
@@ -1070,7 +1071,9 @@ size_t strlcat(char *dst, const char *src, size_t siz)
 
     return dlen + slen + strlen(src + slen); /* count does not include NUL */
 }
+#endif
 
+#ifndef HAVE_STRLCPY
 /* strlcpy based on OpenBSDs strlcpy */
 /*
  * Copy src to string dst of size siz.  At most siz-1 characters
@@ -1096,3 +1099,4 @@ size_t strlcpy(char *dst, const char *src, size_t siz)
 
     return slen + strlen(src + slen); /* count does not include NUL */
 }
+#endif
